c++/src/kj/compat/http.c++

```diff
@@ -5063,20 +5056,23 @@ private:
 };
 
 HttpServer::HttpServer(kj::Timer& timer, HttpHeaderTable& requestHeaderTable, HttpService& service,
-                       Settings settings)
-    : HttpServer(timer, requestHeaderTable, &service, settings,
+                       Settings settings, kj::Maybe<HttpServerErrorHandler&> errorHandler)
+    : HttpServer(timer, requestHeaderTable, &service, settings, errorHandler,
                  kj::newPromiseAndFulfiller<void>()) {}
 
 HttpServer::HttpServer(kj::Timer& timer, HttpHeaderTable& requestHeaderTable,
-                       HttpServiceFactory serviceFactory, Settings settings)
-    : HttpServer(timer, requestHeaderTable, kj::mv(serviceFactory), settings,
+                       HttpServiceFactory serviceFactory, Settings settings,
+                       kj::Maybe<HttpServerErrorHandler&> errorHandler)
+    : HttpServer(timer, requestHeaderTable, kj::mv(serviceFactory), settings, errorHandler,
                  kj::newPromiseAndFulfiller<void>()) {}
 
 HttpServer::HttpServer(kj::Timer& timer, HttpHeaderTable& requestHeaderTable,
                        kj::OneOf<HttpService*, HttpServiceFactory> service,
-                       Settings settings, kj::PromiseFulfillerPair<void> paf)
+                       Settings settings, kj::Maybe<HttpServerErrorHandler&> errorHandler,
+                       kj::PromiseFulfillerPair<void> paf)
     : timer(timer), requestHeaderTable(requestHeaderTable), service(kj::mv(service)),
-      settings(settings), onDrain(paf.promise.fork()), drainFulfiller(kj::mv(paf.fulfiller)),
+      settings(settings), errorHandler(errorHandler.orDefault(*this)),
```

Comment:
Interesting... I hadn't seen this pattern of defaulting to a privately inherited default implementation before.  I guess it's used here because it simplifies the ownership in the default case?

Reply:
Yes, exactly. I wouldn't espouse this as a pattern, but I'm lazy and it seemed to make sense. I should probably make `HttpServerErrorHandler` pure abstract anyway, though...